EKF/estimator_interface.cpp

```diff
@@ -404,21 +404,56 @@ void EstimatorInterface::setOpticalFlowData(uint64_t time_usec, flow_message *fl
 
 			// NOTE: the EKF uses the reverse sign convention to the flow sensor. EKF assumes positive LOS rate is produced by a RH rotation of the image about the sensor axis.
 			// copy the optical and gyro measured delta angles
-			optflow_sample_new.gyroXYZ = - flow->gyrodata;
+
+			bool no_gyro = false;
+			imuSample matching_imu_sample;
+			Vector3f matching_gyro_sample;
+
+			/* for optical flow boards with no gyro. The driver will set the gyro x, y and z fields in the optical_flow message to NAN so that  
+			 * we can check this here and get the gyro data with the matching timestamp from the imu buffer. The gyro data is divided by dt to 
+			 * convert it into a rate before removing it from the flow to remove the rotation component. This is done to prevent errors when the flow
+			 * and gyro data are integrated over different times */
+			if (!ISFINITE(flow->gyrodata(0)) && !ISFINITE(flow->gyrodata(1)) && !ISFINITE(flow->gyrodata(2))) {
+				no_gyro = true;
+				_imu_buffer.read_first_older_than(optflow_sample_new.time_us, &matching_imu_sample);
```

Comment:
This introduces point sampling of  gyro data and with flow sensors that output at a lower rate, eg 10Hz which is undesirable as it can introduce significant errors due to aliasing of body angular motioin. If we are going to do the reconstruction of flow->gyrodata inside the EKF, then it should be across the integration time interval of the flow sensor. 

That means we need a method of sampling and accumulating the gyro data between optical flow measurements.

Reply:
I will have a look to see how we can achieve this cleanly using the high rate IMU data. IMU data  is accumulated and downsampled to 8msec before prior to being written to the EKF's internal buffer, so we will need to run the accumulation for optical flow as a separate function called from inside the setIMUData function before the downsampling.